File: sample-data-repository/src/test/java/nl/haploid/octowight/sample/data/CaptainResourceFactoryTest.java
package nl.haploid.octowight.sample.data;

import mockit.Injectable;
import mockit.Mocked;
import mockit.StrictExpectations;
import mockit.Tested;
import nl.haploid.octowight.registry.data.ResourceRoot;
import nl.haploid.octowight.sample.TestData;
import nl.haploid.octowight.sample.repository.PersonDmo;
import nl.haploid.octowight.sample.repository.RoleDmo;
import nl.haploid.octowight.sample.repository.RoleDmoRepository;
import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class CaptainResourceFactoryTest {

	@Tested
	private CaptainResourceFactory captainResourceFactory;

	@Injectable
	private RoleDmoRepository roleDmoRepository;

	@Test
	public void testFromResourceRoot(final @Mocked RoleDmo roleDmo) {
		final ResourceRoot resourceRoot = TestData.resourceRoot();
		final PersonDmo personDmo = TestData.personDmo();
		new StrictExpectations() {{
			roleDmoRepository.findOne(resourceRoot.getAtomId());
			times = 1;
			result = roleDmo;
			roleDmo.setAtomOrigin(resourceRoot.getAtomOrigin());
			times = 1;
			roleDmo.getPerson();
			times = 1;
			result = personDmo;
		}};
		final CaptainResource captainResource = captainResourceFactory.fromResourceRoot(resourceRoot);
		assertEquals(resourceRoot.getResourceId(), captainResource.getId());
		assertEquals(2, captainResource.getAtoms().size());
		assertEquals(resourceRoot.getVersion(), captainResource.getVersion());
	}
}

File: sample-data-repository/src/test/scala/nl/haploid/octowight/sample/data/CaptainResourceFactoryTest.scala
package nl.haploid.octowight.sample.data

import nl.haploid.octowight.registry.data.Atom
import nl.haploid.octowight.sample.repository.{RoleDmo, RoleDmoRepository}
import nl.haploid.octowight.sample.{AbstractTest, TestData}
import nl.haploid.octowight.{Mocked, Tested}

class CaptainResourceFactoryTest extends AbstractTest {
  @Tested private[this] val captainResourceFactory: CaptainResourceFactory = null
  @Mocked private[this] val roleDmoRepository: RoleDmoRepository = null

  "Captain resource factory" should "build from a resource root" in {
    val roleDmo = mock[RoleDmo]
    val resourceRoot = TestData.resourceRoot
    val personDmo = TestData.personDmo
    val roleAtom = mock[Atom]
    expecting {
      roleDmoRepository.findOne(resourceRoot.root.id) andReturn roleDmo once()
      roleDmo.setOrigin(resourceRoot.root.origin) once()
      roleDmo.getPerson andReturn personDmo once()
      roleDmo.toAtom andReturn roleAtom once()
    }
    whenExecuting(roleDmo, roleDmoRepository) {
      val captainResource = captainResourceFactory.fromResourceRoot(resourceRoot).getOrElse(fail())
      captainResource.id should be(resourceRoot.resourceId)
      captainResource.atoms should have size 2
      captainResource.version should be(resourceRoot.version)
      captainResource.model.id should be(personDmo.getId)
      captainResource.model.name should be(personDmo.getName)
    }
  }
}
